app/src/main/java/org/p2p/wallet/striga/signup/interactor/StrigaSignupInteractor.kt

```diff
@@ -0,0 +1,119 @@
+import org.p2p.wallet.striga.signup.repository.model.StrigaSignupDataType
+import org.p2p.wallet.striga.signup.validation.StrigaSignupDataValidator
+import org.p2p.wallet.utils.unsafeLazy
+
+typealias ValidationResult = Pair<Boolean, List<StrigaSignupFieldState>>
+
+class StrigaSignupInteractor(
+    private val appScope: AppScope,
+    private val validator: StrigaSignupDataValidator,
+    private val countryRepository: CountryRepository,
+    private val signupDataRepository: StrigaSignupDataLocalRepository
+) {
+
+    private val signupDataCache = mutableMapOf<StrigaSignupDataType, StrigaSignupData>()
+    private var needsSave = false
+
+    private val firstStepDataTypes: Set<StrigaSignupDataType> by unsafeLazy {
+        setOf(
+            StrigaSignupDataType.EMAIL,
+            StrigaSignupDataType.PHONE_NUMBER,
+            StrigaSignupDataType.FIRST_NAME,
+            StrigaSignupDataType.LAST_NAME,
+            StrigaSignupDataType.DATE_OF_BIRTH,
+            StrigaSignupDataType.COUNTRY_OF_BIRTH
+        )
+    }
+
+    private val secondStepDataTypes: Set<StrigaSignupDataType> by unsafeLazy {
+        setOf(
+            StrigaSignupDataType.OCCUPATION,
+            StrigaSignupDataType.SOURCE_OF_FUNDS,
+            StrigaSignupDataType.COUNTRY,
+            StrigaSignupDataType.CITY,
+            StrigaSignupDataType.CITY_ADDRESS_LINE,
+            StrigaSignupDataType.CITY_POSTAL_CODE,
+            StrigaSignupDataType.CITY_STATE
+        )
+    }
+
+    fun notifyDataChanged(type: StrigaSignupDataType, newValue: String) {
+        signupDataCache[type] = StrigaSignupData(type, newValue)
+        needsSave = true
+    }
+
+    fun validateFirstStep(): ValidationResult {
+        return validateStep(firstStepDataTypes)
+    }
+
+    fun validateSecondStep(): ValidationResult {
+        return validateStep(secondStepDataTypes)
+    }
+
+    suspend fun getSelectedCountry(): Country {
+        // todo: get saved country
+        return countryRepository.detectCountryOrDefault()
+    }
+
+    suspend fun findPhoneMaskByCountry(country: Country): String? {
+        return countryRepository.findPhoneMaskByCountry(country)
+    }
+
+    suspend fun findCountryByNameCode(countryNameCode: String?): Country? {
+        if (countryNameCode.isNullOrBlank()) return null
+        return countryRepository.findCountryByNameCode(countryNameCode)
+    }
+
+    suspend fun getSignupData(): List<StrigaSignupData> {
+        return when (val data = signupDataRepository.getUserSignupData()) {
+            is StrigaDataLayerResult.Success -> {
+                data.value
+            }
+            is StrigaDataLayerResult.Failure -> {
+                Timber.e(data.error, "Striga signup data: failed to get")
+                emptyList()
+            }
+        }
+    }
+
+    /**
+     * This method saves data only if it was changed
+     */
+    fun saveChanges() {
+        if (needsSave) {
+            needsSave = false
```

Comment:
i think we can update everytime.
DB transaction is not so complex to update again. 

On the otherside we can easily break this logic and I don't really like checks like boolean value cache in interactor